### PR TITLE
feat(snap): Add snap packaging

### PR DESCRIPTION
**Pull request by farshidtz**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features): https://github.com/edgexfoundry/edgex-docs/pull/790

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-examples/blob/master/.github/CONTRIBUTING.md.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports
Are there any new imports or modules? If so, what are they used for and why?

## Specific Instructions

Building locally is possible using Snapcraft 7:
```bash
snapcraft  -v
```
You can skip this and install the pre-built snap for this PR.

```bash
# install platform
# We need the 2.2.0+2 which includes the add-on service in secret store
# For older versions add manually: https://docs.edgexfoundry.org/2.2/getting-started/Ch-GettingStartedSnapUsers/#add-on-services
snap install edgexfoundry

# install pre-built snap for this PR
snap install edgex-device-usb-camera --channel=edge/pr-30

# if built manually, connect the edgex-secretstore-token interface to obtain a token
snap connect edgexfoundry:edgex-secretstore-token edgex-device-usb-camera:edgex-secretstore-token

# connect the camera interface :
snap connect edgex-device-usb-camera:camera :camera
```

Configure and start:
```bash
sudo mv /var/snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/general.usb.camera.toml.example \
        /var/snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/general.usb.camera.toml

# make necessary changes (e.g. the device path)
sudo nano /var/snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/general.usb.camera.toml

# start
# this starts both the rstp server and device-usb-camera apps
snap start edgex-device-usb-camera

# get the logs
snap logs -f edgex-device-usb-camera
```

Start streaming:
```bash
curl -X PUT -d '{
    "StartStreaming": {
      "InputImageSize": "640x480",
      "OutputVideoQuality": "5"
    }
}' http://localhost:59882/api/v2/device/name/example-camera/StartStreaming

# check the stream
mplayer rtsp://localhost:8554/stream/example-camera

# stop streaming (to release your camera)
curl -X PUT -d '{}' http://localhost:59882/api/v2/device/name/example-camera/StopStreaming
```

## Other information

The snap is 130MB+. It may be possible to reduce the size by removing unused shared library object files pulled in as ffmpeg dependencies.

### CODE REVIEW

**Comment by farshidtz** [comment]
@MonicaisHer please review. 
There are a few more todos though:
- [ ] Add the service as addon-service 
- [ ] Fetch appropriate metadata
- [ ] Transfer snap ownership to Canonical

File: .gitignore
Comment (tonyespy) [inline .gitignore]: Missing NL

File: cmd/res/devices/general.usb.camera.toml.example
Comment (tonyespy) [inline cmd/res/devices/general.usb.camera.toml.example]: Missing NL

I think this is because you merged master into your branch vs. using `git pull -r` to rebase your patches? While your approach works, it means there's more non-related changes to review...

File: snap/README.md
Comment (tonyespy) [inline snap/README.md]: Why add a hidden comment?

File: snap/README.md
Comment (tonyespy) [inline snap/README.md:28]: This actually just connects the interface.

In a development scenario, the developer would need to configure the additional settings for the service, then restart the edgexfoundry snap's security-secretstore-setup service.

File: snap/README.md
Comment (tonyespy) [inline snap/README.md]: Maybe note why this is needed, namely that the snapd `camera` interface grants permission for the service to access all USB camera on the system.

Comment (tonyespy) [reply snap/README.md]: Also we should submit a request to the store team to grant this interface to the snap automatically.

File: snap/README.md
Comment (tonyespy) [inline snap/README.md]: Is there a reason to think that some of the included shared libraries are unused?

File: snap/local/build-helpers/include/videodev2.h
Comment (tonyespy) [inline snap/local/build-helpers/include/videodev2.h]: Why are we copying a Linux kernel header file into the snap build-helpers directory?

```
dpkg -S videodev2.h
linux-hwe-5.13-headers-5.13.0-40: /usr/src/linux-hwe-5.13-headers-5.13.0-40/include/uapi/linux/videodev2.h
linux-hwe-5.13-headers-5.13.0-41: /usr/src/linux-hwe-5.13-headers-5.13.0-41/include/uapi/linux/videodev2.h
linux-hwe-5.11-headers-5.11.0-34: /usr/src/linux-hwe-5.11-headers-5.11.0-34/include/uapi/linux/videodev2.h
linux-hwe-5.13-headers-5.13.0-40: /usr/src/linux-hwe-5.13-headers-5.13.0-40/include/linux/videodev2.h
linux-libc-dev:amd64: /usr/include/linux/videodev2.h
linux-hwe-5.13-headers-5.13.0-41: /usr/src/linux-hwe-5.13-headers-5.13.0-41/include/linux/videodev2.h
linux-hwe-5.11-headers-5.11.0-34: /usr/src/linux-hwe-5.11-headers-5.11.0-34/include/linux/videodev2.h
linux-hwe-5.8-headers-5.8.0-63: /usr/src/linux-hwe-5.8-headers-5.8.0-63/include/linux/videodev2.h
linux-hwe-5.8-headers-5.8.0-63: /usr/src/linux-hwe-5.8-headers-5.8.0-63/include/uapi/linux/videodev2.h
```

**Note** - the above output also shows this header included the dev package for glibc. I honestly don't know enough about v4l2 to say why.

In general, it's not a good idea to copy header files from other projects. If you do, then it's also best practice to be very clear about where the file was sourced from and why. This file is also dual licensed (GPLv2 or BSD), and has some terms about copyright notices that must be included in both **source** and **binary** distributions.

I also ran a diff of a few of the different versions of the header file and they're not all the same across different kernel versions.

This also brings up questions as to whether this device service is agnostic to the underlying kernel version?

File: snap/README.md
Comment (tonyespy) [inline snap/README.md]: Is it possible to run this snap in non-secure mode?

File: snap/local/hooks/cmd/install/install.go
Comment (tonyespy) [inline snap/local/hooks/cmd/install/install.go]: Wouldn't it be better to use wildcards? Otherwise, we'd need to update this hook everytime upstream adds new profiles.

File: snap/snapcraft.yaml
Comment (tonyespy) [inline snap/snapcraft.yaml:3]: Do the `summary` and `description` also come from the `metadata` part?

File: snap/snapcraft.yaml
Comment (tonyespy) [inline snap/snapcraft.yaml]: While the camera service itself is licensed as Apache 2.0, some of the other components in this snap have other licenses (e.g. rtsp-simple-server is MIT).

Also this snap is including ffmpeg which is licensed under LGPLv2, however some components of ffmpeg are licensed under GPLv2 (see https://www.ffmpeg.org/legal.html). Was this reviewed as part of the new service review?

Same question about libpulse0 (which again is licensed under LGPLv2)? The pulseaudio license also mentions that if D-Bus support is enabled, the PulseAudio client library (libpulse) MAY need to be licensed under the GPL, depending on the license adopted for libdbus. Also do we know why service requires libpulse0? Is there a runtime requirement for the PulseAudio server in order to use this service?

File: snap/snapcraft.yaml
Comment (tonyespy) [inline snap/snapcraft.yaml]: No NL

File: snap/snapcraft.yaml
Comment (tonyespy) [inline snap/snapcraft.yaml]: The existing header file from where? I checked my snapcraft lxc build container and the only `videodev2.h` file I could find is the one you're copying here.

**Review by tonyespy — COMMENTED** [review]
Please see my inline comments, there are quite a few open questions which need answers.

File: snap/README.md
Comment (MonicaisHer) [reply snap/README.md]: yes, it is possible, I have tested its streaming function in no-secure mode:
```
snap set edgex-device-usb-camera app-options=true
snap set edgex-device-usb-camera config.edgex-security-secret-store=false
```

File: snap/README.md
Comment (MonicaisHer) [inline snap/README.md]:
```
sudo snap set edgexfoundry app-options=true
```
this is needed

File: snap/README.md
Comment (MonicaisHer) [inline snap/README.md]: could you add `sudo` for these commands?

File: cmd/res/devices/general.usb.camera.toml.example
Comment (farshidtz) [reply cmd/res/devices/general.usb.camera.toml.example]: This PR is adding that example.
A clean merge of main into branch does not affect the diff, because the diff is calculated between the branch and main. On the other hand, rebasing will require a re-review of everything because there is no guaranry that the author hasn't changed something else.

File: snap/README.md
Comment (farshidtz) [reply snap/README.md:28]: This is explained above. Will improve the text to make it clearer.

File: snap/local/hooks/cmd/install/install.go
Comment (MonicaisHer) [inline snap/local/hooks/cmd/install/install.go]: could we add a pre-defined example device into `/var/snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/`?

File: snap/README.md
Comment (farshidtz) [reply snap/README.md]: I intend to uncomment it when the auto connection is made possible, i.e. after transferring ownership to Canonical.

File: snap/README.md
Comment (farshidtz) [reply snap/README.md]: > Also we should submit a request to the store team to grant this interface to the snap automatically.

Right. That's why I said "currently not automatically connected".
Will add a note to clarify.

File: snap/README.md
Comment (farshidtz) [reply snap/README.md]: This is a draft note (in draft PR). "It may be possible to reduce" and this needs further investigation. Some libraries may not be needed. A statically built go binary is another way to remove reliance on those.

File: snap/local/build-helpers/include/videodev2.h
Comment (farshidtz) [reply snap/local/build-helpers/include/videodev2.h]: I've explained in the snapcraft file why the header is needed.

Good point about the license. However, the header file is included in the resulting snap whether or not I package this version of it. The snap is simply replacing the existing one with a copy that has the set of definitions needed for the compilation.

Regarding being agnostic to the underlying kernel, this probably an issue from the dependency: https://github.com/vladimirvivien/go4vl#compilation-requirements

File: snap/snapcraft.yaml
Comment (farshidtz) [reply snap/snapcraft.yaml:3]: yes

**Review by MonicaisHer — DISMISSED** [review]
@farshidtz Thanks a lot for the PR. I have tested it with secure and non-secure modes. Both work as expected for video streaming with my camera Aukey-PC-LM1E.

File: snap/local/hooks/cmd/install/install.go
Comment (farshidtz) [reply snap/local/hooks/cmd/install/install.go]: The example device is currently provided under the readonly snap files. The user can copy and modify it if needed:
```
sudo cp /snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/general.usb.camera.toml.example \
    /var/snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/general.usb.camera.toml
```
I will think about it. Maybe both the example general profile and device can be copied to writable directories but with `.example` suffix. Otherwise, they will get uploaded to core-metadata even if not used or relevant.

File: cmd/res/devices/general.usb.camera.toml.example
Comment (farshidtz) [reply cmd/res/devices/general.usb.camera.toml.example]: Created a separate PR for the example. Will remove from this PR once merged:
- https://github.com/edgexfoundry/device-usb-camera/pull/33

File: snap/README.md
Comment (farshidtz) [reply snap/README.md]: Added. Note that this section will be removed once this PR is merged and the snaps released:
- https://github.com/edgexfoundry/edgex-go/pull/4034

File: snap/snapcraft.yaml
Comment (farshidtz) [reply snap/snapcraft.yaml]: > do we know why service requires libpulse0? Is there a runtime requirement for the PulseAudio server in order to use this service?

It is needed by `pprobe`. Added a comment.

Edit: turns out that libpulseaudio it was already there, just that it was missing from path. Removed it as staged package.

File: snap/README.md
Comment (farshidtz) [reply snap/README.md]: Added a link to camera interface docs and note to clarify that auto connection will happen in future.

File: snap/local/hooks/cmd/install/install.go
Comment (farshidtz) [reply snap/local/hooks/cmd/install/install.go]: All example profiles are copied to the read only snap files and available to the user. The code here is just copying the general one to snap data. 

We may actually be better off not copying any profile/device to snap data because they would be uploaded to core metadata even if not desired. See my other comment: https://github.com/edgexfoundry/device-usb-camera/pull/30#discussion_r879114979

File: snap/snapcraft.yaml
Comment (farshidtz) [reply snap/snapcraft.yaml]: > While the camera service itself is licensed as Apache 2.0, some of the other components in this snap have other licenses (e.g. rtsp-simple-server is MIT).

This should be fine since MIT is permissive.
I've also considered not including rtsp server in the snap and let users install their own server (e.g. [v4l2-rtspserver](https://snapcraft.io/v4l2-rtspserver)). But in the end, I decided to keep it since it is only 9MB and disabled by default. The docker image includes it as well.

> Also this snap is including ffmpeg which is licensed under LGPLv2, however some components of ffmpeg are licensed under GPLv2 (see https://www.ffmpeg.org/legal.html). Was this reviewed as part of the new service review?


I did not review the docker base images and packages, not sure if someone else did. @iain-anderson @lenny-intel an idea?
This would be a project issue and not specific to the snap.

File: snap/snapcraft.yaml
Comment (farshidtz) [reply snap/snapcraft.yaml]: Not sure; probably from core20 (package linux-headers-5.4.0-110).

The go4vl library mentions v5.10.x as the compilation dependency. See https://github.com/vladimirvivien/go4vl#compilation-requirements

Comment (farshidtz) [reply snap/snapcraft.yaml]: Should no longer be an issue when we upgrade to core22.

File: snap/snapcraft.yaml
Comment (jpwhitemn) [reply snap/snapcraft.yaml]: According to explicit section in GNU's license agreement (section 5 in the document linked below), we are ok to use the library and still distribute as Apache 2.  https://www.gnu.org/licenses/old-licenses/lgpl-2.1.html.  This is known as the linking exception.  The same was applied when we worked with 0MQ.  LF legal cannot provide legal advise (strange but true), but their reading of this also pointed us to this exception.  (see https://en.wikipedia.org/wiki/GPL_linking_exception).

File: snap/README.md
Comment (farshidtz) [reply snap/README.md]: Moved to docs:
- https://github.com/edgexfoundry/edgex-docs/pull/772

File: snap/README.md
Comment (farshidtz) [reply snap/README.md]: Moved to PR description.

File: cmd/res/devices/general.usb.camera.toml.example
Comment (farshidtz) [reply cmd/res/devices/general.usb.camera.toml.example]: PR has been merged. Removed the example from this PR and merged main into the branch.

File: snap/snapcraft.yaml
Comment (tonyespy) [reply snap/snapcraft.yaml]: @jpwhitemn you're misinterpreting my comment. I acknowledged that the ffmpeg package is licensed as **LGPLv2**, however there are optional parts of ffmpeg that are licensed as **GPLv2**. My question was simply did we review our usage of ffmpeg to ensure we didn't use one of these optional parts?

File: snap/snapcraft.yaml
Comment (farshidtz) [reply snap/snapcraft.yaml]: Switched to core22 and removed the sideloaded header.

File: snap/snapcraft.yaml
Comment (iain-anderson) [reply snap/snapcraft.yaml]: As far as I can tell the ffmpeg support doesn't link the ffmpeg library, it runs an ffmpeg command as a separate process?

File: snap/local/build-helpers/include/videodev2.h
Comment (farshidtz) [reply snap/local/build-helpers/include/videodev2.h]: See issues #37

File: snap/local/hooks/cmd/install/install.go
Comment (farshidtz) [reply snap/local/hooks/cmd/install/install.go]: Added code to copy the example device to writable directory. I kept the `.example` suffix to avoid unintentional loading. It isn't needed when dynamic discovery is used.

File: snap/local/hooks/cmd/install/install.go
Comment (farshidtz) [reply snap/local/hooks/cmd/install/install.go]: Added code to copy the example device to writable directory. I kept the `.example` suffix to avoid unintentional loading. It isn't needed when dynamic discovery is used.

**Review by MonicaisHer — DISMISSED** [review]
@farshidtz  Thanks for the changes. The newest edgex-device-usb-camera snap works as expected.

File: snap/snapcraft.yaml
Comment (MonicaisHer) [inline snap/snapcraft.yaml:15]: The `content` can be removed here, as it defaults to the slot name.

File: snap/local/hooks/cmd/install/install.go
Comment (tonyespy) [inline snap/local/hooks/cmd/install/install.go]: Why are we copying an example device file when it won't actually be imported unless someone manually renames the file?

**Review by tonyespy — COMMENTED** [review]
I had one comment inline about copying an example device file into` $SNAP_DATA`.

The bigger issue is that I can't seem to build the snap anymore either natively on 22.04 using `--destructive-mode` or using `--use-lxd` on 20.04. In both cases I'm using snapcraft 7.0.2 from `latest/candidate`.

File: snap/local/hooks/cmd/install/install.go
Comment (farshidtz) [reply snap/local/hooks/cmd/install/install.go]: This is to maintain consistency among different device services, where configuration files are available in the snap data directories (examples or not). It will be hard for a user to find the example general device if it is not copied over to snap data.

**Comment by farshidtz** [comment]
> The bigger issue is that I can't seem to build the snap anymore either natively on 22.04 using `--destructive-mode` or using `--use-lxd` on 20.04. In both cases I'm using snapcraft 7.0.2 from `latest/candidate`.

It builds fine with snapcraft 6 (current latest/stable)
As I've reported [here](https://forum.snapcraft.io/t/call-for-testing-snapcraft-7/30201/35?u=farshidtz), there is a bug in snapcraft 7.0.2 preventing the loading of summary and description from the AppStream metadata file. Should be fixed in the next patch release in a few days. Will put this PR on hold until that is done, because snapcraft 7 is the tool that officially supports the core22 base used in this snap. It may be do additional validations worth waiting for.

**Comment by farshidtz** [comment]
We can now build the snap using snapcraft 7.0.4 but there are some unresolved issues in regards to packaging of the hooks and licenses.

**Comment by farshidtz** [comment]
Snapcraft 7.0.8 packages the licenses but the hooks issue isn't resolved yet. I have reported it [here](https://bugs.launchpad.net/snapcraft/+bug/1980089).

In the meantime, I've added a workaround (https://github.com/edgexfoundry/device-usb-camera/pull/30/commits/4eccf07e03728cf6a18b800588ce5f506dea6fca) which solves the issue for this snap.

Another thing I noticed when building was an added 15MB to the size of the snap. It looks like many additional files are being packaged automatically from dependencies (e.g. under usr/share apart from docs). We'll have to dig deeper and remove them in future PRs.

File: snap/local/helper-go/go.sum
Comment (MonicaisHer) [inline snap/local/helper-go/go.sum]: `go mod tidy`?

**Review by MonicaisHer — DISMISSED** [review]
@farshidtz Thanks for the hook refactor. 
I have tested this snap. .env files and config files are in place. Everything works as expected.

**Review by MonicaisHer — DISMISSED** [review]

**Comment by farshidtz** [comment]
Found and fixed a bug while working on the [tests](https://github.com/canonical/edgex-snap-testing/pull/75).

The tests don't run on this PR yet because Github Actions get enabled for the first time after this PR has merged. A sample run can be reviewed from my fork: https://github.com/farshidtz/edgex-device-usb-camera/runs/7136890125

File: snap/README.md
Comment (tonyespy) [inline snap/README.md:15]: We should mention that building this snap remains "challenging" and at least offer some options. I don't think the given option "just works".

File: snap/README.md
Comment (tonyespy) [inline snap/README.md:36]: Have we requested these permissions yet?

File: snap/hooks/configure
Comment (tonyespy) [inline snap/hooks/configure:2]: This is a workaround for snapcraft 7, correct? Did you ever get a response from the snapcraft team on this? At minimum, we should probably add a note to this script noting that it's a workaround.

File: snap/hooks/install
Comment (tonyespy) [inline snap/hooks/install:3]: same as configure comment

File: snap/snapcraft.yaml
Comment (tonyespy) [reply snap/snapcraft.yaml]: My point is that the snap metadata is saying the license of this snap is "Apache 2.0", and that's not technically correct, as the snap (and most likely our docker image) contains more than just the device service Go binary. The guidance from the snapcraft.io [reference](https://snapcraft.io/docs/snapcraft-yaml-reference) for `snapcraft.yaml `say this needs to be an SPDX license expression (see https://spdx.dev/spdx-specification-21-web-version/), which in this case should be a list of licenses as there's doesn't seem to be an option for "multiple licenses" anymore.



For instance, here are a number of shared libraries that and up in the snap and their associated licenses:

- libpango-1.0.0 - LGPL2
- libvorbisenc2 - BSD
- libxvidcore4 - GPL2+

And also while reviewing the files in the snap, I don't see a license file for the rtsp-server itself.

**Review by tonyespy — CHANGES_REQUESTED** [review]
While I able to build with snapcraft 6 natively on a 22.04 machine, things still fail for me if I try to build natively using snapcraft 7.0.8 on the same machine. If you want the details, I can provide.

I was able to successfully install the snap, however the testing instructions weren't sufficient for me to actually test the snap.
  - any suggestions on what camera I should use?
  - what are the changes I need to make to the device yaml file?

Also note, there are still some issues with licensing (please see inline comment).

File: snap/README.md
Comment (farshidtz) [reply snap/README.md:36]: Not yet. We will do so once we have a stable source and build that can be reviewed by the store.

**Comment by farshidtz** [comment]
@tonyespy Thanks for the review.

> While I able to build with snapcraft 6 natively on a 22.04 machine, things still fail for me if I try to build natively using snapcraft 7.0.8 on the same machine. If you want the details, I can provide.
> 

The snap build and tests (added in this PR) pass on Github (https://github.com/farshidtz/edgex-device-usb-camera/runs/7136890125; see comments above) and Launchpad (https://launchpad.net/~canonical-edgex/+snap/draft-edgex-device-usb-camera). Also builds fine on my machine (Ubuntu 22.04) and @MonicaisHer's. I'm not sure what issue you have faced. Maybe a clean build will resolve it?


> I was able to successfully install the snap, however the testing instructions weren't sufficient for me to actually test the snap.
> 
> * any suggestions on what camera I should use?
> * what are the changes I need to make to the device yaml file?

This PR is just adding snap packaging. Please refer to the README: https://github.com/edgexfoundry/device-usb-camera#tested-devices
There is a PR adding a few more to the list: https://github.com/edgexfoundry/device-usb-camera/pull/49

Changes may be required to the local [device path](https://github.com/edgexfoundry/device-usb-camera/blob/73afea3ace466341e80d38fe07fac6720914120d/cmd/res/devices/general.usb.camera.toml.example#L8) if you have more than one connected USB camera.

File: snap/hooks/configure
Comment (farshidtz) [reply snap/hooks/configure:2]: As mentioned in [this comment](https://github.com/edgexfoundry/device-usb-camera/pull/30#issuecomment-1168729209), I've reported the bug.

We are no longer affected by it because we have refactored the hooks to have a single binary executable. This reduces the build complexity (code, time) and snap size. The same improvement is being rolled out to other snaps.

File: snap/snapcraft.yaml
Comment (farshidtz) [reply snap/snapcraft.yaml]: There are close to 200 packages pulled in for ffmpeg and libzmq5. Many of those (incl. libzmq ones) have multiple licenses. It is simply not practical to identify and include all these, and maintain them over time. I don't see any other Docker image or snap listing all licenses.

I found some useful information at:
- https://www.linuxfoundation.org/tools/docker-containers-what-are-the-open-source-licensing-considerations/
- https://opensource.stackexchange.com/a/8393

I believe the license in the Dockerfile is the Dockerfile's license, rather than the license of Docker images built with it. For the snap, the license field appear to be the license of the resulting snap packages. So, IMO, it is most appropriate to remove the optional `license` field. All licenses remain available under snap's `/usr/share/doc` directory.

The license for rtsp-simple-server is at snap's `/usr/share/doc/rtsp-simple-server/LICENSE`.

File: snap/README.md
Comment (farshidtz) [reply snap/README.md:15]: As explained in [another comment](https://github.com/edgexfoundry/device-usb-camera/pull/30#issuecomment-1175016386), we have been able to build it without problems locally and on CI servers. If there are any OS-specific challenges related to using snapcraft to build core22-based snaps, it should be raised on snapcraft issue trackers.

File: snap/snapcraft.yaml
Comment (farshidtz) [reply snap/snapcraft.yaml]: I have removed the license field from snapcraft.yaml: https://github.com/edgexfoundry/device-usb-camera/pull/30/commits/4c4ab6af09efedeff12eed0a9f83e12c8148dad6

File: snap/snapcraft.yaml
Comment (farshidtz) [reply snap/snapcraft.yaml:15]: Sure. But let's just leave it in for now.

**Comment by tonyespy** [comment]
> @tonyespy Thanks for the review.
> 
> > While I able to build with snapcraft 6 natively on a 22.04 machine, things still fail for me if I try to build natively using snapcraft 7.0.8 on the same machine. If you want the details, I can provide.
> 
> The snap build and tests (added in this PR) pass on Github (https://github.com/farshidtz/edgex-device-usb-camera/runs/7136890125; see comments above) and Launchpad (https://launchpad.net/~canonical-edgex/+snap/draft-edgex-device-usb-camera). Also builds fine on my machine (Ubuntu 22.04) and @MonicaisHer's. I'm not sure what issue you have faced. Maybe a clean build will resolve it?

It definitely failed for me using snapcraft 7.0.8 on Jammy, but it's now working with 7.0.9, so not worth investigating any further.

> > I was able to successfully install the snap, however the testing instructions weren't sufficient for me to actually test the snap.
> > 
> > * any suggestions on what camera I should use?
> > * what are the changes I need to make to the device yaml file?
> 
> This PR is just adding snap packaging. Please refer to the README: https://github.com/edgexfoundry/device-usb-camera#tested-devices There is a PR adding a few more to the list: #49

I was reacting to the fact that you provided some instructions, but then in the middle left out a bunch of details. I'm reviewing a PR not the service. While not required, detailed test instructions make the process of reviewing easier for the reviewers. So that said, I'm not going to test anything more than installing the snap.
 
> Changes may be required to the local [device path](https://github.com/edgexfoundry/device-usb-camera/blob/73afea3ace466341e80d38fe07fac6720914120d/cmd/res/devices/general.usb.camera.toml.example#L8) if you have more than one connected USB camera.

**Review by tonyespy — APPROVED** [review]
Based on our conversation earlier today, I'm approving this PR "as is".

**Review by MonicaisHer — APPROVED** [review]

**Comment by farshidtz** [comment]
Thanks for the reviews. This PR is ready to merge, leaving the following as future work:
- Request auto connection permissions for the `camera` interface
- Add snap options to selectively set autostart apps
- Set snap's license based on all the bundled packages

@cloudxxx8 could you please squash and merge? Please remove all the commits history from squash commit description.

**Review by lenny-goodell — APPROVED** [review]

**Pull request merged**
